File: plugins/OrchidForms/src/main/java/com/eden/orchid/presentations/FormsGenerator.java
package com.eden.orchid.presentations;

import com.eden.orchid.api.OrchidContext;
import com.eden.orchid.api.generators.OrchidGenerator;
import com.eden.orchid.api.options.annotations.Description;
import com.eden.orchid.api.options.annotations.Option;
import com.eden.orchid.api.options.annotations.StringDefault;
import com.eden.orchid.api.theme.pages.OrchidPage;
import com.eden.orchid.utilities.OrchidUtils;
import org.json.JSONObject;

import javax.inject.Inject;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Description("Indexes form definitions so they can be easily referenced from components on different pages.")
public class FormsGenerator extends OrchidGenerator {

    private final FormsModel model;

    @Option("baseDir")
    @StringDefault("forms")
    public String formsBaseDir;

    @Inject
    public FormsGenerator(OrchidContext context, FormsModel model) {
        super(context, "forms", 20);
        this.model = model;
    }

    @Override
    public List<? extends OrchidPage> startIndexing() {
        Map<String, Form> forms = new HashMap<>();

        JSONObject formsData = context.getDatafiles(OrchidUtils.normalizePath(formsBaseDir));

        for(String key : formsData.keySet()) {
            forms.put(key, new Form(context, key, formsData.getJSONObject(key)));
        }

        model.initialize(forms);

        return null;
    }

    @Override
    public void startGeneration(List<? extends OrchidPage> pages) {

    }

}

File: plugins/OrchidForms/src/main/kotlin/com/eden/orchid/forms/FormsGenerator.kt
package com.eden.orchid.forms

import com.eden.orchid.api.OrchidContext
import com.eden.orchid.api.generators.OrchidGenerator
import com.eden.orchid.api.options.annotations.Description
import com.eden.orchid.api.options.annotations.Option
import com.eden.orchid.api.options.annotations.StringDefault
import com.eden.orchid.forms.model.Form
import com.eden.orchid.forms.model.FormsModel
import com.eden.orchid.utilities.OrchidUtils

@Description("Indexes form definitions so they can be easily referenced from components on different pages.",
        name = "Forms"
)
class FormsGenerator : OrchidGenerator<FormsModel>(GENERATOR_KEY, Stage.CONTENT) {

    companion object {
        const val GENERATOR_KEY = "forms"
    }

    @Option
    @StringDefault("forms")
    @Description("The base directory in local resources to look for forms in.")
    lateinit var baseDir: String

    override fun startIndexing(context: OrchidContext): FormsModel {
        val forms = getFormsByDatafiles(context)

        return FormsModel(forms)
    }

    private fun getFormsByDatafiles(context: OrchidContext) : List<Form> {
        return context
            .getResourceEntries(
                OrchidUtils.normalizePath(baseDir),
                context.parserExtensions.toTypedArray(),
                false,
                null
            )
            .map { resource ->
                resource.reference.isUsePrettyUrl = false
                val fileData = context.parse(resource.reference.extension, resource.content)
                val key = resource.reference.originalFileName
                Form(context, key, fileData)
            }
    }

}
